### PR TITLE
[Retiarii] Grid search, random and evolution strategy

### PR DESCRIPTION
**Pull request by ultmaster**
This PR refines and adds new strategies to Retiarii.

- [x] Grid search
- [x] Random
- [x] Evolution
- [x] tests

### CODE REVIEW

File: nni/retiarii/strategy/bruteforce.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            submit_models(get_targeted_model(base_model, applied_mutators, sample))
+
+
+class _RandomSampler(Sampler):
+    def choice(self, candidates, mutator, model, index):
+        return random.choice(candidates)
+
+
+class Random(BaseStrategy):
+    def __init__(self, variational=False, dedup=True):
```
Comment (QuanluZhang) [inline nni/retiarii/strategy/bruteforce.py:64]: what is the meaning of `variational`?

Comment (QuanluZhang) [reply nni/retiarii/strategy/bruteforce.py:64]: got it...

File: nni/retiarii/strategy/bruteforce.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                    for mutator in applied_mutators:
+                        model = mutator.apply(model)
+                    _logger.info('New model created. Applied mutators are: %s', str(applied_mutators))
+                    submit_models(model)
+                else:
+                    time.sleep(self._polling_interval)
+        else:
+            _logger.info('Random search running in fixed size mode. Dedup: %s.', 'on' if self.dedup else 'off')
+            search_space = dry_run_for_search_space(base_model, applied_mutators)
+            for sample in random_generator(search_space, dedup=self.dedup):
```
Comment (QuanluZhang) [inline nni/retiarii/strategy/bruteforce.py:91]: what will webui show if strategy exits but maxtrialnum and maxduration are not reached?

Comment (ultmaster) [reply nni/retiarii/strategy/bruteforce.py:111]: I tested. The whole experiment directly exits and dispatcher (strategy) is terminated.

File: nni/retiarii/strategy/bruteforce.py (hunk 1 above)
Comment (ultmaster) [reply nni/retiarii/strategy/bruteforce.py:84]: Docs added.

File: nni/retiarii/strategy/bruteforce.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/retiarii/strategy/bruteforce.py:111]: dispatcher is not strategy, why dispatcher exits?...

File: nni/retiarii/strategy/evolution.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    Follows "Algorithm 1" in Real et al. "Regularized Evolution for Image Classifier Architecture Search".
+
+    Parameters
+    ----------
+    optimize_mode : str
+        Can be one of "maximize" and "minimize".
+    population_size : int
+        The number of individuals to keep in the population.
+    cycles : int
+        The number of cycles (trials) the algorithm should run for.
```
Comment (QuanluZhang) [inline nni/retiarii/strategy/evolution.py:38]: cycle is a little misleading, cycle means trial in the paper?

File: nni/retiarii/strategy/bruteforce.py (hunk 2 above)
Comment (ultmaster) [reply nni/retiarii/strategy/bruteforce.py:111]: "Dispatcher terminated" is printed on the console. Afterwards, no more trials appear on the WebUI. I'm not sure about the details.

File: nni/retiarii/strategy/evolution.py (hunk 3 above)
Comment (ultmaster) [reply nni/retiarii/strategy/evolution.py:38]: The terminology is used in paper and their open-source implementation.

image

Comment (QuanluZhang) [reply nni/retiarii/strategy/evolution.py]: ok

File: nni/retiarii/strategy/bruteforce.py (hunk 2 above)
Comment (QuanluZhang) [reply nni/retiarii/strategy/bruteforce.py:111]: got it, this is a normal behavior

**Review by QuanluZhang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

**Pull request merged**